compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -1587,23 +1587,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             self.ty_to_value_string(rcvr_ty.peel_refs())
         };
         if let SelfSource::MethodCall(_) = source {
-            let first_arg = if let Some(CandidateSource::Impl(impl_did)) = static_candidates.get(0)
-                && let Some(assoc) = self.associated_value(*impl_did, item_name)
-                && assoc.kind == ty::AssocKind::Fn
-            {
+            let first_arg = static_candidates.get(0).and_then(|candidate_source| {
+                let (assoc_did, impl_ty) = match candidate_source {
```

Comment:
```suggestion
                let (assoc_did, self_ty) = match candidate_source {
```

“`impl_ty`” doesn't feel right for `CandidateSource::Trait`.